packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -467,18 +412,43 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
   } else if (stateP.controlState === 'UPDATE_SOURCE_MAPPINGS_PROPERTIES') {
     const res = resW as ExcludeRetryableEsError<ResponseType<typeof stateP.controlState>>;
     if (Either.isRight(res)) {
-      return {
-        ...stateP,
-        controlState: 'CLEANUP_UNKNOWN_AND_EXCLUDED',
-        targetIndex: stateP.sourceIndex.value!, // We preserve the same index, source == target (E.g: ".xx8.7.0_001")
-        versionIndexReadyActions: Option.none,
-      };
+      switch (res.right) {
+        case UpdateSourceMappingsPropertiesResult.Compatible:
+          return {
+            ...stateP,
+            controlState: 'CLEANUP_UNKNOWN_AND_EXCLUDED',
+          };
+        case UpdateSourceMappingsPropertiesResult.Incompatible:
+          return {
+            ...stateP,
+            controlState: 'CHECK_UNKNOWN_DOCUMENTS',
+          };
+        case UpdateSourceMappingsPropertiesResult.Updated:
+          return {
+            ...stateP,
+            // Skip to 'OUTDATED_DOCUMENTS_SEARCH_OPEN_PIT' so that if a new plugin was
+            // installed / enabled we can transform any old documents and update
+            // the mappings for this plugin's types.
+            controlState: 'OUTDATED_DOCUMENTS_SEARCH_OPEN_PIT',
+            // Source is a none because we didn't do any migration from a source index
+            sourceIndex: Option.none,
+            targetIndex: stateP.sourceIndex.value,
+            // in this scenario, a .kibana_X.Y.Z_001 index exists that matches the current kibana version
+            // aka we are NOT upgrading to a newer version
+            // we inject the source index's current mappings in the state, to check them later
+            targetIndexMappings: mergeMigrationMappingPropertyHashes(
+              stateP.targetIndexMappings,
+              stateP.sourceIndexMappings.value
+            ),
+          };
+      }
```

Comment:
I like that this makes the possible state transitions really dense and explicit, we can see what condition triggers what state transition :+1: